### PR TITLE
Support multiple IPv6Token 'static' addreses on an interface

### PR DESCRIPTION
**Pull request by kpfleming**
(this builds on #14831 and will be rebased once that is merged)

This patch allows multiple addresses using 'static' IPv6Tokens to be generated for a single network interface.

This is my first time using the systemd data structures (Set, in this case), so I won't be surprised if I've done something incorrectly (especially in the area of memory management). Feedback is very much welcome :-) A Set is used here instead of a List to handle the case where the supplied configuration produces the same address more than once; using a Set will ensure that `address_configure` is only called once for each produced address, avoiding an unnecessary error.

### CODE REVIEW

**Comment by kpfleming** [comment]
I haven't been able to find any documentation on how to execute the systemd-networkd test suite; if anyone can point me to that documentation or provide guidance, I'd like to ensure that the tests in this branch pass as they should.

File: src/network/networkd-ndisc.c [hunk 1]
```diff
@@ -274,45 +288,74 @@ static int ndisc_router_generate_address(Link *link, unsigned prefixlen, uint32_
...
+                        new_address->prefixlen = prefixlen;
+                        new_address->flags = IFA_F_NOPREFIXROUTE|IFA_F_MANAGETEMPADDR;
+                        new_address->cinfo.ifa_prefered = lifetime_preferred;
+
+                        r = set_put(addresses, new_address);
+                        if (r < 0)
+                                return r;
+                } else {
+                        address_free(new_address);
+                }
```
Comment (yuwata) [inline src/network/networkd-ndisc.c:314]: Please drop `{}` for single line block.

File: src/network/networkd-ndisc.c [hunk 2]
```diff
@@ -274,45 +288,74 @@ static int ndisc_router_generate_address(Link *link, unsigned prefixlen, uint32_
...
 
         return 0;
 }
+
 static int ndisc_router_process_autonomous_prefix(Link *link, sd_ndisc_router *rt) {
         uint32_t lifetime_valid, lifetime_preferred, lifetime_remaining;
+        _cleanup_set_free_free_ Set *addresses = NULL;
         _cleanup_(address_freep) Address *address = NULL;
         Address *existing_address;
+        Address *a;
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: `Address *existing_address, *a;`

File: src/network/networkd-ndisc.c [hunk 3]
```diff
@@ -274,45 +288,74 @@ static int ndisc_router_generate_address(Link *link, unsigned prefixlen, uint32_
...
+        Iterator i;
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Move this to before `int r;`

File: src/network/networkd-ndisc.c [hunk 4]
```diff
@@ -349,34 +392,37 @@ static int ndisc_router_process_autonomous_prefix(Link *link, sd_ndisc_router *r
...
+                r = address_configure(a, link, ndisc_netlink_address_message_handler, true);
+                if (r < 0) {
+                        log_link_warning_errno(link, r, "Could not set SLAAC address: %m");
+                        link_enter_failed(link);
+                        return r;
+                }
         }
+
         if (r > 0)
                 link->ndisc_messages++;
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: This must be in the loop.

**Comment by yuwata** [comment]
The CI failure should be fixed by #14836.

**Comment by yuwata** [comment]
rebased.

**Comment by yuwata** [comment]
@kpfleming I've dropped Singed-off-by thing in the commit message, as we do not require that.

File: src/network/networkd-ndisc.c [hunk 5]
```diff
@@ -254,62 +254,108 @@ static int ndisc_router_process_default(Link *link, sd_ndisc_router *rt) {
...
-        if (!prefix) {
+                if (have_address) {
+                        new_address->prefixlen = prefixlen;
+                        new_address->flags = IFA_F_NOPREFIXROUTE|IFA_F_MANAGETEMPADDR;
+                        new_address->cinfo.ifa_prefered = lifetime_preferred;
+
+                        r = set_put(addresses, new_address);
+                        if (r < 0)
+                                return r;
+                } else
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: `TAKE_PTR(new_address);` must be added after `set_put()`.

File: src/network/networkd-ndisc.c [hunk 6]
```diff
@@ -254,62 +254,108 @@ static int ndisc_router_process_default(Link *link, sd_ndisc_router *rt) {
...
+                new_address->in_addr.in6.s6_addr[14] = link->mac.ether_addr_octet[4];
+                new_address->in_addr.in6.s6_addr[15] = link->mac.ether_addr_octet[5];
+                new_address->prefixlen = prefixlen;
+                new_address->flags = IFA_F_NOPREFIXROUTE|IFA_F_MANAGETEMPADDR;
+                new_address->cinfo.ifa_prefered = lifetime_preferred;
+
+                r = set_put(addresses, new_address);
+                if (r < 0)
+                        return r;
         }
```
Comment (yuwata) [inline src/network/networkd-ndisc.c:345]: Same here.

**Comment by yuwata** [comment]
I fixed the use-after-free errors, and updated the log messages. I hope the CIs will pass.

**Comment by ssahani** [comment]
Why this is called static ? can you point me in the RFC. And for variable renaming a  patch. Please squash those.

**Comment by yuwata** [comment]
@ssahani Please read https://github.com/systemd/systemd/pull/14415#discussion_r375400334.

**Comment by ssahani** [comment]
That does not mean we call it static.

**Comment by yuwata** [comment]
Do you have any suggestion?

**Comment by ssahani** [comment]
Need to see what others use. We just can't adopt anything randomly. I will read up rfc again.

**Comment by kpfleming** [comment]
I am not aware of any RFC which documents the 'IPv6 token' behavior of the Linux kernel, which is what is being emulated here. If you can find a better name than 'static' for it that's good with me but I just didn't want it to be called 'eui64' since that does have a meaning :-)

**Comment by ssahani** [comment]
Can we just can call `token` simply to avoid any introducing new term.

**Comment by kpfleming** [comment]
The variable name is already IPv6Token, so you're suggesting it would be:

`IPv6Token=token:::1a:2b:3c:4d`

Is that correct?

**Comment by ssahani** [comment]
`IPv6Token=token:::1a:2b:3c:4d` yes but it introduces a `static` term and no where it's used. but we don' t want to introduce a new term. With the backward compatibility `IPv6Token=::1a:2b:3c:4d`  would work .  Maybe we can allow `static`. I am not too sure about it though.

**Comment by kpfleming** [comment]
Realistically, we should convert this to a configuration section instead of a variable, but that's a much larger change. Examples:

```
[IPv6SLAAC]
mode=eui64
```

```
[IPv6SLAAC]
mode=fixed
interface_identifier=::1a:2b:3c:4d
```

```
[IPv6SLAAC]
mode=rfc7217
prefix=2001:db8::/64
hash_ifname=true/false
hash_mac=true/false
secret_key=<string>
```

**Comment by ssahani** [comment]
Yes initial idea was that but we found `:` thing makes very shorter .

```
hash_ifname=true/false
hash_mac=true/false
secret_key=<string>
```
With this we automated and above all this are required .

**Comment by ssahani** [comment]
I think we can merge this. I like the documents. Very well written. Thank you

**Pull request merged**